src/main/java/bio/terra/workspace/service/datareference/utils/DataReferenceValidationUtils.java

```diff
@@ -1,56 +1,67 @@
 import java.util.regex.Pattern;
 import org.apache.commons.lang3.StringUtils;
 import org.springframework.stereotype.Component;
 
+/** A collection of validation functions for data references. */
 @Component
 public class DataReferenceValidationUtils {
 
   private DataRepoService dataRepoService;
-  private final Pattern nameValidationPattern = Pattern.compile("^[a-zA-Z0-9][_a-zA-Z0-9]{0,62}$");
 
-  public DataReferenceValidationUtils(DataRepoService dataRepoService) {
-    this.dataRepoService = dataRepoService;
-  }
+  /**
+   * Names must be 1-63 characters long, and may consist of alphanumeric characters and underscores
+   * (but may not start with an underscore). These restrictions match TDR snapshot name
+   * restrictions.
+   */
+  public static final Pattern NAME_VALIDATION_PATTERN =
+      Pattern.compile("^[a-zA-Z0-9][_a-zA-Z0-9]{0,62}$");
 
-  public void validateReferenceName(String name) {
-    if (StringUtils.isEmpty(name) || !nameValidationPattern.matcher(name).matches()) {
+  public static void validateReferenceName(String name) {
+    if (StringUtils.isEmpty(name) || !NAME_VALIDATION_PATTERN.matcher(name).matches()) {
       throw new InvalidDataReferenceException(
           "Invalid reference name specified. Name must be 1 to 63 alphanumeric characters or underscores, and cannot start with an underscore.");
     }
   }
 
-  public DataRepoSnapshot validateReference(
-      ReferenceTypeEnum referenceType,
-      DataRepoSnapshot reference,
+  public DataReferenceValidationUtils(DataRepoService dataRepoService) {
+    this.dataRepoService = dataRepoService;
+  }
+
+  /**
+   * Validates a referenceObject, with specific validation rules varying based on the actual type of
+   * the object.
+   */
+  public void validateReferenceObject(
+      DataReferenceType referenceType,
+      ReferenceObject reference,
       AuthenticatedUserRequest userReq) {
 
     switch (referenceType) {
       case DATA_REPO_SNAPSHOT:
-        validateDataRepoReference(reference, userReq);
-        return reference;
+        validateSnapshotReference((SnapshotReference) reference, userReq);
+        return;
       default:
         throw new InvalidDataReferenceException(
             "Invalid reference type specified. Valid types include: "
-                + DATA_REPO_SNAPSHOT.toString());
+                + DataReferenceType.DATA_REPO_SNAPSHOT.toString());
     }
   }
 
-  private void validateDataRepoReference(DataRepoSnapshot ref, AuthenticatedUserRequest userReq) {
-    if (StringUtils.isBlank(ref.getInstanceName()) || StringUtils.isBlank(ref.getSnapshot())) {
+  private void validateSnapshotReference(SnapshotReference ref, AuthenticatedUserRequest userReq) {
+    if (StringUtils.isBlank(ref.instanceName()) || StringUtils.isBlank(ref.snapshot())) {
       throw new InvalidDataReferenceException(
           "Invalid Data Repo Snapshot identifier: "
               + "instanceName and snapshot must both be provided.");
```

Comment:
should this be validation on SnapshotReference creation? What would it mean to have a SnapshotReference object that didn't pass this validation?

Reply:
Yeah, this should probably be at object creation time.

Reply:
Sorry, I meant should this happen in the method SnapshotReference.create? Should we restrict bad SnapshotReference objects from even being created, or do we want to just verify at the controller that new incoming requests don't fail this validation. If we don't want them created, we have to think about what to do when we change validation in a non-backwards compatible way i.e. DAO reads will start failing. If we only want to check at the controller, we may not notice if we change validation in a non-backwards compatible way.

Reply:
Ah, gotcha. My concern with moving it to `create` was around deserialization, but agreed that introduces concerns around breaking validation. I think we can mitigate some amount of that with testing